os/arch/arm/src/s5j/s5j_boot.c

```diff
@@ -131,9 +157,30 @@ void up_copyvectorblock(void)
 }
 
 #ifdef CONFIG_ARMV7M_MPU
+static inline unsigned int arm_is_highvec(void)
+{
+	unsigned int vec = 0;
+	__asm__ __volatile__
+	(
+		"\tmrc p15, 0, %0, c1, c0, 0"
+		: "=r"(vec)
+		:
+		: "memory"
+	);
+
+	return (vec & SCTLR_V);
+}
+
 int s5j_mpu_initialize(void)
 {
-#ifdef CONFIG_ARCH_CHIP_S5JT200
+#if defined(CONFIG_BUILD_PROTECTED)
+	int i;
+
+	for (i = 0; i < (sizeof(regions_info) / sizeof(struct mpu_region_info)); i++) {
+		regions_info[i].call(regions_info[i].base, regions_info[i].size);
```

Comment:
I guess, This is good approach to assign the MPU regions to different memory address instead of using mpu_allocregion() , Because mpu_allocregion()  just uses  global variable and assigns the mpu region without any intelligence.   
I recommend using this way of assigning the MPU regions even for FLAT build as well.  This enhances the runtime assignment of MPU regions to different memory address.   One best example is, MPU_STACKGUARD feature requires MPU region to be assigned at runtime , This can't be achieved using the existing API like mpu_allocregion() .

Reply:
As of now changes for FLAT build not required (as discussed internally).